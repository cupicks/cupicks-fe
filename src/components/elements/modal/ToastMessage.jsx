import { useState } from "react";
import styled from "styled-components";

const ToastMessage = (props) => {
  const {
    text = "",
    imageUrl = "",
    timer = 2000,
    backgroundShadow = false,
    smallFont = false,
  } = props;
  const [modalShow, setModalShow] = useState(true);

  const backgroundColor = backgroundShadow ? "rgba(0, 0, 0, 0.3)" : "";
  const fontSize = smallFont ? "1.4rem" : "1.8rem";

  setTimeout(() => {
    setModalShow(false);
  }, timer);

  return (
    <>
      {modalShow && (
        <StModal
          fontSize={fontSize}
          backgroundColor={backgroundColor}
          timer={timer * 0.001}
          onClick={() => {
            setModalShow(false);
          }}
        >
          <div className="contents">
            {imageUrl && (
              <img
                className="illust"
                src={imageUrl}
                alt="일러스트레이션 이미지"
              />
            )}

            {text}
          </div>
        </StModal>
      )}
    </>
  );
};

export default ToastMessage;

const StModal = styled.div`
  width: 100vw;
  height: 100vh;

  padding: 0 10px;

  display: flex;
  flex-flow: column;
  justify-content: center;
  align-items: center;

  position: fixed;
  top: 0;
  left: 0;
  z-index: 999999;

  font-size: 18px;
  font-weight: 600;
  line-height: 150%;
  text-align: center;
  word-wrap: break-word;

  background-color: ${(props) => props.backgroundColor};
  animation: fadeInOut ${(props) => props.timer}s forwards;

  @keyframes fadeInOut {
    0% {
      opacity: 0;
    }
    10% {
      opacity: 1;
    }
    90% {
      opacity: 1;
    }
    100% {
      opacity: 0;
    }
  }

  .contents {
    display: flex;
    flex-flow: column;
    align-items: center;
    gap: 20px;

    width: 90%;
    max-width: 350px;
    border-radius: 15px;

    padding: 35px 0;

    position: relative;
    transform: translateY(-150px);

    background-color: #fff;
    color: #393939;
    border: 1px solid #393939;
    box-shadow: 0px 4px 26px rgba(0, 0, 0, 0.25);

    font-size: ${(props) => props.fontSize};
    font-weight: 700;

    box-sizing: initial;
    white-space: pre-wrap;
  }

  .button_close {
    position: absolute;
    top: 11px;
    right: 15px;

    cursor: pointer;
  }

  button {
    all: unset;
    padding: 8px 60px;

    background: #101010;
    color: #fff;
    border-radius: 10px;

    cursor: pointer;
  }

  img.illust {
    width: 80%;
  }
`;
