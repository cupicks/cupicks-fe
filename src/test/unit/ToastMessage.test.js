import ToastMessage from "../../components/elements/modal/ToastMessage.jsx";

import {
  render,
  screen,
  cleanup,
  fireEvent,
  renderHook,
} from "@testing-library/react";
import renderer from "react-test-renderer";

// afterEach(cleanup);

describe("ToastMessage component", () => {
  // arrange
  const propsText = "사용자가 입력하는 메시지";
  const propsTimer = 2000;

  const component = renderer.create(
    <ToastMessage
      text={propsText}
      timer={propsText}
      imageUrl="/image.png"
      backgroundShadow={false}
      smallFont={false}
      title="modal"
    />,
  );

  // act
  // 가짜 타이머
  // setModalShow(false);

  // assert
  it("ToastMessage가 render되었습니다.", () => {
    expect(component).toBeInTheDocument;
    expect(component).toBeVisible;
  });

  it("ToastMessage가 기존 스냅샷과 같습니다.", () => {
    expect(component).toMatchSnapshot();
  });

  it("ToastMessage의 modal이 화면에 보입니다.", () => {
    // const modal = screen.findByTitle("modal");
    const modal = screen.findByTitle("modal");
    expect(modal).toBeInTheDocument;
    expect(modal).toBeTruthy;
  });

  it("ToastMessage의 modal이 화면에 없습니다.", () => {
    const modal = screen.queryByTitle("modal");
    expect(modal).not.toBeInTheDocument;
  });

  it("전달받은 메시지가 화면에 출력됩니다.", () => {
    const modal = screen.findByTestId("modal");
    expect(propsText).toBeInTheDocument;
  });

  it("props로 전달받은 이미지가 화면에 출력됩니다.", () => {
    // const { imageUrl } = component.props;
    const imageElement = screen.findByRole("img");
    expect(imageElement).toBeInTheDocument;
    // src 출력에 에러가 없음
    // src에 우리 주소 포함
    // expect(imageElement.src).toMatch(imageUrl);
  });

  it("modalShow state이 true입니다.", () => {
    // const { result, renderer } = renderHook(({ modalShow } = {}) => modalShow, {
    //   initialProps: { modalShow: false, text: "토스트 메시지" },
    // });
    // expect(result.current).toEqual(false);
    // const modal = screen.findByRole("aside");
    // expect(modal).toBeInTheDocument;
  });

  it("modalShow값이 false일 때, StModal 컴포넌트를 render하지 않음", () => {
    // state값을 확인해야 함
    // const modal = screen.getByRole("aside", { exact: false });
    const modal = document.querySelector(".contents");
    expect(modal).toBeInTheDocument;
  });

  it("timer시간 후에 modalShow값이 false로 바뀜", () => {
    const modal = screen.findByRole(
      "aside",
      { exact: false },
      { timeout: 2000 },
    );
    expect(modal).not.toBeInTheDocument;
  });

  it("modalShow값이 false로 바뀌었음", () => {});
});
