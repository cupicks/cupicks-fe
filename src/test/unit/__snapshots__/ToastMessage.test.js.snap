// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastMessage component ToastMessage가 기존 스냅샷과 같습니다. 1`] = `<div />`;
